### ISSUE TITLE
Implement HostInfo for Haiku

### ISSUE THREAD

**Issue by kenz-gelsoft** [kenz-gelsoft/llvm-project#3]
part of #1 

```
/boot/home/src/llvm-project/lldb/source/Interpreter/CommandInterpreter.cpp: In member function 'void lldb_private::CommandInterpreter::SourceInitFileHome(lldb_private::CommandReturnObject&, bool)':
/boot/home/src/llvm-project/lldb/source/Interpreter/CommandInterpreter.cpp:2198:19: error: 'GetProgramFileSpec' is not a member of 'lldb_private::HostInfo' {aka 'lldb_private::HostInfoPosix'}
 2198 |         HostInfo::GetProgramFileSpec().GetFilename().GetStringRef();
      |                   ^~~~~~~~~~~~~~~~~~
```
HostInfo implementation required per OS.
```
~/src/llvm-project> grep -R GetProgramFileSpec lldb/
lldb/bindings/interface/SBHostOS.i:    GetProgramFileSpec ();
lldb/docs/python_api.rst:  :skip: SBHostOS_GetProgramFileSpec
lldb/include/lldb/API/SBHostOS.h:  static lldb::SBFileSpec GetProgramFileSpec();
lldb/include/lldb/Host/freebsd/HostInfoFreeBSD.h:  static FileSpec GetProgramFileSpec();
lldb/include/lldb/Host/linux/HostInfoLinux.h:  static FileSpec GetProgramFileSpec();
lldb/include/lldb/Host/macosx/HostInfoMacOSX.h:  static FileSpec GetProgramFileSpec();
lldb/include/lldb/Host/netbsd/HostInfoNetBSD.h:  static FileSpec GetProgramFileSpec();
lldb/include/lldb/Host/openbsd/HostInfoOpenBSD.h:  static FileSpec GetProgramFileSpec();
lldb/include/lldb/Host/windows/HostInfoWindows.h:  static FileSpec GetProgramFileSpec();
lldb/source/API/SBHostOS.cpp:SBFileSpec SBHostOS::GetProgramFileSpec() {
lldb/source/API/SBHostOS.cpp:                                    GetProgramFileSpec);
lldb/source/API/SBHostOS.cpp:  sb_filespec.SetFileSpec(HostInfo::GetProgramFileSpec());
lldb/source/API/SBHostOS.cpp:  LLDB_REGISTER_STATIC_METHOD(lldb::SBFileSpec, SBHostOS, GetProgramFileSpec,
lldb/source/Host/freebsd/HostInfoFreeBSD.cpp:FileSpec HostInfoFreeBSD::GetProgramFileSpec() {
lldb/source/Host/linux/HostInfoLinux.cpp:FileSpec HostInfoLinux::GetProgramFileSpec() {
lldb/source/Host/linux/HostInfoLinux.cpp:  file_spec.GetDirectory() = GetProgramFileSpec().GetDirectory();
lldb/source/Host/macosx/objcxx/HostInfoMacOSX.mm:FileSpec HostInfoMacOSX::GetProgramFileSpec() {
lldb/source/Host/netbsd/HostInfoNetBSD.cpp:FileSpec HostInfoNetBSD::GetProgramFileSpec() {
lldb/source/Host/openbsd/HostInfoOpenBSD.cpp:FileSpec HostInfoOpenBSD::GetProgramFileSpec() {
lldb/source/Host/windows/HostInfoWindows.cpp:FileSpec HostInfoWindows::GetProgramFileSpec() {
lldb/source/Interpreter/CommandInterpreter.cpp:        HostInfo::GetProgramFileSpec().GetFilename().GetStringRef();
```

**Comment by kenz-gelsoft** [comment]
BSDs has almost identical header
```diff
~/src/llvm-project/lldb/include/lldb/Host> diff -u freebsd/HostInfoFreeBSD.h netbsd/HostInfoNetBSD.h 
--- freebsd/HostInfoFreeBSD.h   2023-06-17 16:12:44.781189120 +0900
+++ netbsd/HostInfoNetBSD.h     2023-06-17 16:12:44.789315584 +0900
@@ -1,4 +1,4 @@
-//===-- HostInfoFreeBSD.h ---------------------------------------*- C++ -*-===//
+//===-- HostInfoNetBSD.h ---------------------------------------*- C++ -*-===//
 //
 // Part of the LLVM Project, under the Apache License v2.0 with LLVM Exceptions.
 // See https://llvm.org/LICENSE.txt for license information.
@@ -6,8 +6,8 @@
 //
 //===----------------------------------------------------------------------===//
 
-#ifndef lldb_Host_freebsd_HostInfoFreeBSD_h_
-#define lldb_Host_freebsd_HostInfoFreeBSD_h_
+#ifndef lldb_Host_netbsd_HostInfoNetBSD_h_
+#define lldb_Host_netbsd_HostInfoNetBSD_h_
 
 #include "lldb/Host/posix/HostInfoPosix.h"
 #include "lldb/Utility/FileSpec.h"
@@ -15,7 +15,7 @@
 
 namespace lldb_private {
 
-class HostInfoFreeBSD : public HostInfoPosix {
+class HostInfoNetBSD : public HostInfoPosix {
 public:
   static llvm::VersionTuple GetOSVersion();
   static bool GetOSBuildString(std::string &s);
```
```diff
~/src/llvm-project/lldb/include/lldb/Host> diff -u freebsd/HostInfoFreeBSD.h openbsd/HostInfoOpenBSD.h 
--- freebsd/HostInfoFreeBSD.h   2023-06-17 16:12:44.781189120 +0900
+++ openbsd/HostInfoOpenBSD.h   2023-06-17 16:12:44.790626304 +0900
@@ -1,4 +1,4 @@
-//===-- HostInfoFreeBSD.h ---------------------------------------*- C++ -*-===//
+//===-- HostInfoOpenBSD.h --------------------------------------*- C++ -*-===//
 //
 // Part of the LLVM Project, under the Apache License v2.0 with LLVM Exceptions.
 // See https://llvm.org/LICENSE.txt for license information.
@@ -6,8 +6,8 @@
 //
 //===----------------------------------------------------------------------===//
 
-#ifndef lldb_Host_freebsd_HostInfoFreeBSD_h_
-#define lldb_Host_freebsd_HostInfoFreeBSD_h_
+#ifndef lldb_Host_openbsd_HostInfoOpenBSD_h_
+#define lldb_Host_openbsd_HostInfoOpenBSD_h_
 
 #include "lldb/Host/posix/HostInfoPosix.h"
 #include "lldb/Utility/FileSpec.h"
@@ -15,7 +15,7 @@
 
 namespace lldb_private {
 
-class HostInfoFreeBSD : public HostInfoPosix {
+class HostInfoOpenBSD : public HostInfoPosix {
 public:
   static llvm::VersionTuple GetOSVersion();
   static bool GetOSBuildString(std::string &s);
```

**Comment by kenz-gelsoft** [comment]
Linux has extension or expose more impelementation detail
```diff
~/src/llvm-project/lldb/include/lldb/Host> diff -u freebsd/HostInfoFreeBSD.h linux/HostInfoLinux.h 
--- freebsd/HostInfoFreeBSD.h   2023-06-17 16:12:44.781189120 +0900
+++ linux/HostInfoLinux.h       2023-06-17 16:12:44.783286272 +0900
@@ -1,4 +1,4 @@
-//===-- HostInfoFreeBSD.h ---------------------------------------*- C++ -*-===//
+//===-- HostInfoLinux.h -----------------------------------------*- C++ -*-===//
 //
 // Part of the LLVM Project, under the Apache License v2.0 with LLVM Exceptions.
 // See https://llvm.org/LICENSE.txt for license information.
@@ -6,21 +6,41 @@
 //
 //===----------------------------------------------------------------------===//
 
-#ifndef lldb_Host_freebsd_HostInfoFreeBSD_h_
-#define lldb_Host_freebsd_HostInfoFreeBSD_h_
+#ifndef lldb_Host_linux_HostInfoLinux_h_
+#define lldb_Host_linux_HostInfoLinux_h_
 
 #include "lldb/Host/posix/HostInfoPosix.h"
 #include "lldb/Utility/FileSpec.h"
+#include "llvm/ADT/StringRef.h"
 #include "llvm/Support/VersionTuple.h"
 
+#include <string>
+
 namespace lldb_private {
 
-class HostInfoFreeBSD : public HostInfoPosix {
+class HostInfoLinux : public HostInfoPosix {
+  friend class HostInfoBase;
+
+private:
+  // Static class, unconstructable.
+  HostInfoLinux();
+  ~HostInfoLinux();
+
 public:
+  static void Initialize();
+
   static llvm::VersionTuple GetOSVersion();
   static bool GetOSBuildString(std::string &s);
   static bool GetOSKernelDescription(std::string &s);
+  static llvm::StringRef GetDistributionId();
   static FileSpec GetProgramFileSpec();
+
+protected:
+  static bool ComputeSupportExeDirectory(FileSpec &file_spec);
+  static bool ComputeSystemPluginsDirectory(FileSpec &file_spec);
+  static bool ComputeUserPluginsDirectory(FileSpec &file_spec);
+  static void ComputeHostArchitectureSupport(ArchSpec &arch_32,
+                                             ArchSpec &arch_64);
 };
 }
```

**Comment by kenz-gelsoft** [comment]
GetOSVersion(), GetOSBuildString(), GetOSKernelDescription()
* BSDs parse uname()
* Linux has more consideration for distros

GetProgramFileSpec()
* Linux uses procfs
* BSDs use sysctl() with `KERN_PROC_PATHNAME` for pid
* OpenBSD has empty implementation?
  ```c++
  FileSpec HostInfoOpenBSD::GetProgramFileSpec() {
    static FileSpec g_program_filespec;
    return g_program_filespec;
  }
  ```

**Comment by kenz-gelsoft** [comment]
6bf6616

**Comment by kenz-gelsoft** [comment]
abafb13

```c++
bool HostInfoHaiku::GetOSBuildString(std::string &s) {
//  int mib[2] = {CTL_KERN, KERN_OSREV};
//  char osrev_str[12];
//  uint32_t osrev = 0;
//  size_t osrev_len = sizeof(osrev);
//
//  if (::sysctl(mib, 2, &osrev, &osrev_len, NULL, 0) == 0) {
//    ::snprintf(osrev_str, sizeof(osrev_str), "%-8.8u", osrev);
//    s.assign(osrev_str);
//    return true;
//  }

  s.clear();
  return false;
}
```

**Comment by kenz-gelsoft** [comment]
At first, I tried to retrieve the string that AbouSystem uses. But it uses a private API maybe not exported to apps not bundled.

It seems better to extract them from uname output.

**Comment by kenz-gelsoft** [comment]
Copied from https://github.com/kenz-gelsoft/llvm-project/issues/11#issuecomment-1615323651

> A few steps of negitiation proceed, but when client requests `HostInfo` then server exits.
> 
> I inserted some runtime assertions to fail, to know how unimplemented or doubtful part is called.  But lldb-server is launched as child process and AFAIK parent discards its stderr output.
> ```
>  0x112e32522360 ConnectionFileDescriptor::Read()  fd = 6, dst = 0x7fe736aaced0, dst_len = 8192) => 6, error = (null)
>  <   6> read packet: $OK#9a
>  0x0000112E3264AF70 Communication::Write (src = 0x0000112E3262E9F8, src_len = %lu) connection = 13
>  0x112e32522360 ConnectionFileDescriptor::Write (src = 0x112e3262e9f8, src_len = 13)
>  0x112e32522360 ConnectionFileDescriptor::Write(fd = 6, src = 0x112e3262e9f8, src_len = 13) => 13 (error = (null))
>  <  13> send packet: $qHostInfo#9b
>  this = 0x0000112E3264AF70, dst = 0x00007FE736AACB50, dst_len = 8192, timeout = 10000000 us, connection = 0x0000112E32522360
>  this = 0x0000112E32522360, timeout = 10000000 us
>  MonitorChildProcessThreadFunction ::waitpid (pid = 23708, &status, options = 0) => pid = 23708, status = 0x00000001 (EXITED), signal = 0, exit_state = 1
>  ProcessGDBRemote::MonitorDebugserverProcess(process_wp, pid=23708, signo=0 (0x0), exit_status=1)
>  ProcessGDBRemote::MonitorDebugserverProcess(process = 0x112e3264a6a0)
>  Process::SetExitStatus (status=-1 (0xffffffff), description="debugserver died with an exit status of 0x00000001")
> ```

**Comment by kenz-gelsoft** [comment]
https://github.com/kenz-gelsoft/llvm-project/blob/haiku12/lldb/docs/lldb-gdb-remote.txt#L793

**Comment by kenz-gelsoft** [comment]
Implemented in 3a052f8, it seems to work.

```
 <  13> send packet: $qHostInfo#9b
 this = 0x000011C34DBE0F70, dst = 0x00007F54276F99E0, dst_len = 8192, timeout = 10000000 us, connection = 0x000011C34DAB8400
 this = 0x000011C34DAB8400, timeout = 10000000 us
 0x11c34dab8400 ConnectionFileDescriptor::Read()  fd = 4, dst = 0x7f54276f99e0, dst_len = 8192) => 238, error = (null)
 < 238> read packet: $triple:7838365f36342d756e6b6e6f776e2d6861696b75;ptrsize:8;watchpoint_exceptions_received:after;endian:little;os_build:687265763537313236;os_kernel:687265763537313236204a756c20203120323032332030363a32313a3139;hostname:7368726564646572;#d8
 GDBRemoteCommunicationClient::GetHostInfo parsed host architecture as x86_64, triple as x86_64-unknown-haiku from triple text x86_64-unknown-haiku
```

**Comment by kenz-gelsoft** [comment]
I assume this is done.
